src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedInteractiveInitializerMethod.cs

```diff
@@ -237,7 +237,15 @@ internal TypeSymbol ResultType
             {
                 diagnostics.Add(useSiteDiagnostic, NoLocation.Singleton);
             }
-            resultType = compilation.GetTypeByReflectionType(submissionReturnType, diagnostics);
+            // if no explicit return type is set on ScriptCompilationInfo, it will default to
+            // the _host_ corlib's System.Object (typeof(object)) and should never be null.
+            // In this case, redirect the host System.Object to the compilation's target
+            // corlib's System.Object. This allows cross compiling scripts to build on one
+            // corlib and run on another.
+            // cf. https://github.com/dotnet/roslyn/issues/8506
+            resultType = submissionReturnType == typeof(object)
```

Comment:
It sounds like we would want the code to fail if host specified typeof(object) explicitly, but with this change the code would succeed. I would suggest to change the line 
```
var submissionReturnType = compilation.SubmissionReturnType ?? typeof(object);
``` 
to not overwrite null with typeof(object) and base the   submissionReturnType null in this case and base the current logic on whether submissionReturnType is null instead.